### PR TITLE
Ftr: add pass through proxy factory

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
**What this PR does**:
add pass through factory

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
pass through proxy invoker can invoke user defined service, and pass through all data to user defined service. User can wirte  logic to deal with raw invocation data.
It is very useful to dapr-dubbo support. 

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: common/constant/key.go [hunk 1]
```diff
@@ -50,7 +50,8 @@ const (
 	PROTOCOL_KEY             = "protocol"
 	PATH_SEPARATOR           = "/"
 	//DUBBO_KEY                = "dubbo"
-	SSL_ENABLED_KEY = "ssl-enabled"
+	SSL_ENABLED_KEY  = "ssl-enabled"
+	ParameterTypeKey = "parameter-type-names" // ParameterType key used in dapr, to tranfer rsp param type
```
Comment (watermelo) [inline common/constant/key.go:54]: Maybe follow the code`SSL_ENABLED_KEY `.

File: common/proxy/proxy_factory/pass_through.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	result.SetAttachments(invocation.Attachments())
+	url := getProviderURL(pi.GetUrl())
+
+	arguments := invocation.Arguments()
+	srv := common.ServiceMap.GetServiceByServiceKey(url.Protocol, url.ServiceKey())
+
+	var args [][]byte
+	if len(arguments) == 0 {
+		args = [][]byte{}
+	} else {
```
Comment (watermelo) [inline common/proxy/proxy_factory/pass_through.go:91]: We can reduce a `if-else` branch by
```go
if len(arguments) > 0 {
	args = make([][]byte, 0, len(arguments))
	for _, arg := range arguments {
		if v, ok := arg.([]byte); ok {
			args = append(args, v)
		} else {
			result.Err = perrors.New("the param type is not []byte")
			return result
		}
	}
}
```

File: common/proxy/proxy_factory/pass_through.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+				args = append(args, v)
+			} else {
+				result.Err = perrors.New("the param type is not []byte")
+				return result
+			}
+		}
+	}
+	method := srv.Method()["Service"]
+
+	in := []reflect.Value{srv.Rcvr()}
```
Comment (watermelo) [inline common/proxy/proxy_factory/pass_through.go:104]: Init with length `in := make([]reflect.Value, 5)`.

File: common/proxy/proxy_factory/pass_through.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	in = append(in, reflect.ValueOf(args))
+	in = append(in, reflect.ValueOf(invocation.Attachments()))
+
+	returnValues := method.Method().Func.Call(in)
+
+	var retErr interface{}
+	replyv := returnValues[0]
+	retErr = returnValues[1].Interface()
+
+	if retErr != nil {
```
Comment (watermelo) [inline common/proxy/proxy_factory/pass_through.go:116]: Maybe more appropriate.
```go
	if retErr != nil {
		result.SetError(retErr.(error))
                return result
	}
	if replyv.IsValid() && (replyv.Kind() != reflect.Ptr || replyv.Kind() == reflect.Ptr && replyv.Elem().IsValid()) {
		result.SetResult(replyv.Interface())
	}
	return result
```

File: registry/directory/directory.go [hunk 5]
```diff
@@ -187,9 +187,12 @@ func (dir *RegistryDirectory) refreshAllInvokers(events []*registry.ServiceEvent
 		// loop the updateEvents
 		for _, event := range addEvents {
 			logger.Debugf("registry update, result{%s}", event)
-			logger.Infof("selector add service url{%s}", event.Service)
-			// FIXME: routers are built in every address notification?
-			dir.configRouters()
+			if event.Service != nil {
```
Comment (watermelo) [inline registry/directory/directory.go:190]: `if event != nil && event.Service != nil`

File: README.md [hunk 6]
```diff
@@ -177,7 +177,7 @@ If you are willing to do some code contributions and document contributions to [
 
 ## Community
 
-If u want to communicate with our community, pls scan the following  [dubbobo Ding-Ding QR code](https://mmbiz.qpic.cn/mmbiz_jpg/yvBJb5IiafvnHVBdtia30dxA2hKotr9DEckWsZ7aOJcDWDaSVMGwLmYv8GRgIQtqb4C2svicp8nVkMmGy7yKC5tyA/640?wx_fmt=jpeg&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1) or search our commnity DingDing group code 31363295.
+If u want to communicate with our community, pls scan the following dubbobo Ding-Ding QR code or search our commnity DingDing group code 31363295.
```
Comment (AlexStocks) [inline README.md:180]: "Ding-Ding " -> DingDing

File: common/proxy/proxy_factory/pass_through.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	result := &protocol.RPCResult{}
+	result.SetAttachments(invocation.Attachments())
+	url := getProviderURL(pi.GetUrl())
+
+	arguments := invocation.Arguments()
+	srv := common.ServiceMap.GetServiceByServiceKey(url.Protocol, url.ServiceKey())
+
+	var args [][]byte
+	if len(arguments) > 0 {
+		args = make([][]byte, 0, len(arguments))
```
Comment (AlexStocks) [inline common/proxy/proxy_factory/pass_through.go:90]: the same comment as line 93: pls using gxbytes.AcquireBytes instead.

Comment (LaurenceLiZhixin) [reply common/proxy/proxy_factory/pass_through.go:90]: gxbytes.AcquireBytes() is used to get specific make([]byte, 0, size), not to make [][]byte.

File: common/proxy/proxy_factory/pass_through.go (hunk 4 above)
Comment (LaurenceLiZhixin) [reply common/proxy/proxy_factory/pass_through.go:115]: fixed!

File: common/proxy/proxy_factory/pass_through.go (hunk 3 above)
Comment (LaurenceLiZhixin) [reply common/proxy/proxy_factory/pass_through.go]: fixed

File: common/constant/key.go (hunk 1 above)
Comment (LaurenceLiZhixin) [reply common/constant/key.go]: fixed

**Review by AlexStocks — APPROVED** [review]

**Review by wenxuwan — APPROVED** [review]

**Review by zhaoyunxing92 — APPROVED** [review]
approved these changes

**Pull request merged**